feat(RemovableLabel): create RemovableLabel and LabelRemove

karelhala opened the pull request:
<!--
Thanks for your interest in patternfly-react. We appreciate all issues filed and PRs submitted!

Please make sure you're familiar with and follow the instructions in the
contributing guidelines (found in the CONTRIBUTING.md file).

Please fill out the information below to expedite the review and (hopefully)
merge of your pull request!
-->

<!-- What changes are being made? (What issue is being addressed here?) -->
**What**:
Fixes https://github.com/patternfly/patternfly-react/issues/222

<!-- Please provide a link to your fork's Storybook. See README notes on how to do this. -->
**Link to Storybook**:
https://karelhala.github.io/patternfly-react/?selectedKind=Label&selectedStory=Label%20with%20remove&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybooks%2Fstorybook-addon-knobs

**Additional info**
<!-- feel free to add additional comments -->
This PR creates two new components, which can later be used as standalone components for labels.

jeff-phillips-18 commented:
Neither of these classes have anything Filter about them. The FIlterLabel is really just `<Label bsStyle="info">` which probably should have been used in FilterItem.

Might be better to break this out of Filter all together and make it part of Label, add a property there to specify `removable` to include the x button and add an `onRemove` callback. Thoughts?

karelhala commented:
@jeff-phillips-18 good pont! I'll refactor it so it's part of Label instead of Filter.

priley86 commented on src/components/Filter/FilterRemove.js:
```diff
@@ -0,0 +1,31 @@
+import PropTypes from 'prop-types';
+
+const FilterRemove = ({ className, onFilterRemoveClick, title, ...rest }) => {
+  const classes = cx('pficon pficon-close', className);
```
we voted to start using `classNames` in the PF React Community meeting (was a live survey)  ;)

I will open a separate issue on this now...

jeff-phillips-18 commented on src/components/Filter/Filter.stories.js:
```diff
@@ -11,7 +11,9 @@ import {
-  FilterItem
+  FilterItem,
+  FilterLabel,
+  FilterRemove
```
We don't need the FIlterLabel and FilterRemove anymore, right?

karelhala commented:
@priley86 @jeff-phillips-18 can you retest and review it once again?

It's done in a way, that you can either import new component `<DisposableLabel>` with type and onRemoveClick event, or you can import `<Label>` with same properties. If you don't like new component that's fine I can remove it, but it feels kinda better this way.

jeff-phillips-18 commented on src/components/Label/DisposableLabel.js:
```diff
@@ -0,0 +1,29 @@
+DisposableLabel.propTypes = {
+  /** Children nodes */
+  children: PropTypes.node,
+  /** additional filter item classes */
```
comments still reference filter.

jeff-phillips-18 commented on src/components/Label/Label.stories.js:
```diff
@@ -25,3 +26,5 @@ stories.addWithInfo('Label', () => (
   </div>
 ));
+
+stories.addWithInfo('Label with remove', () => <MockLabelRemove />);
```
This give very little information in the story book. When using a Mock component, it's likely you will need to specify the `text` and the `propTables` to include example source and the Prop Types for the components you want to document.

jeff-phillips-18 commented:
The close link button doesn't look right in the non-filter labels. When in the filter they have the Patternfly styling from the toolbar-pf styling:
```
.toolbar-pf-results .label a {
  color: @color-pf-white;
  display: inline-block;
  margin-left: 5px;
}
```
This is not applied when its outside of the toolbar.

karelhala commented:
@jeff-phillips-18 thanks for the hints. I updated the mocks as requested, is that correct? With the styling, it was a bit off so I updated sass and less based on your suggestions. Not sure if that is correct approach tho.

jeff-phillips-18 commented on less/patternfly-react.less:
```diff
@@ -1,3 +1,10 @@
+    display: inline-block;
+    margin-left: 5px;
+  }
+}
```
This should be in  a separate partial file. Please see https://github.com/patternfly/patternfly-react/blob/master/CONTRIBUTING.md#adding-styling-for-your-components.

Please add `-pf` suffix to `remove-button` to avoid conflicts.

karelhala commented:
@jeff-phillips-18 great points, all should be resolved.

priley86 commented:
No issues w/ JS here... just noting the "x" icon does not look centered to me in your Storybook (but it does seem to be in the PF Filter design docs you reference). Did you update the Storybook?
http://www.patternfly.org/pattern-library/forms-and-controls/filter/

jeff-phillips-18 commented:
@priley86 Agreed. The `x` is one pixel too high, this is a Patternfly core issue..

priley86 commented:
@jeff-phillips-18 maybe we can go ahead and merge this and address the design issue in Core later, and just reference that issue?

priley86 commented on src/components/Label/__mocks__/mockLabelExamples.js:
```diff
@@ -0,0 +1,77 @@
+import React from 'react';
```
I think the vote was to start moving these examples into "Stories" instead of "__mocks__" - my apologies for the confusion on this.
